src/core/server/elasticsearch/cluster_client.ts

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Boom from 'boom';
+import { Client } from 'elasticsearch';
+import { get } from 'lodash';
+import { filterHeaders, Headers } from '../http/router';
+import { Logger } from '../logging';
+import {
+  ElasticsearchClientConfig,
+  parseElasticsearchClientConfig,
+} from './elasticsearch_client_config';
+import { ScopedClusterClient } from './scoped_cluster_client';
+
+/**
+ * The set of options that defines how API call should be made and result be
+ * processed.
+ */
+export interface CallAPIOptions {
+  /**
+   * Indicates whether `401 Unauthorized` errors returned from the Elasticsearch API
+   * should be wrapped into `Boom` error instances with properly set `WWW-Authenticate`
+   * header that could have been returned by the API itself. If API didn't specify that
+   * then `Basic realm="Authorization Required"` is used as `WWW-Authenticate`.
+   */
+  wrap401Errors: boolean;
+}
+
+/**
+ * Calls the Elasticsearch API endpoint with the specified parameters.
+ * @param client Raw Elasticsearch JS client instance to use.
+ * @param endpoint Name of the API endpoint to call.
+ * @param clientParams Parameters that will be directly passed to the
+ * Elasticsearch JS client.
+ * @param options Options that affect the way we call the API and process the result.
+ */
+async function callAPI(
+  client: Client,
+  endpoint: string,
+  clientParams: Record<string, unknown> = {},
+  options: CallAPIOptions = { wrap401Errors: true }
+) {
+  const clientPath = endpoint.split('.');
+  const api: any = get(client, clientPath);
```

Comment:
Is there anyway we could get the actual type information out of `elasticsearch-js`? Now that that package is going to be shipping TS definitions, it'd be great if we could leverage them. Seems like some combination of using `keyof` and conditional types could pull this type information out.

Reply:
The new es client module that is under development and provides types is a new project entirely to replace `elasticsearch-js`. We will move to it, but it's types aren't compatible with `elasticsearch-js`.